compiler/rustc_ast/src/util/comments.rs

```diff
@@ -47,55 +46,66 @@ pub fn beautify_doc_string(data: Symbol) -> String {
             j -= 1;
         }
 
-        lines[i..j].to_vec()
+        if i != 0 || j != lines.len() { Some((i, j)) } else { None }
     }
 
-    /// remove a "[ \t]*\*" block from each line, if possible
-    fn horizontal_trim(lines: Vec<String>) -> Vec<String> {
+    /// remove whitespace-only lines from the start/end of lines
+    fn vertical_trim<'a, 'b>(lines: &'a mut [&'b str], i: usize, j: usize) -> &'a mut [&'b str] {
+        &mut lines[i..j]
+    }
+
+    fn get_horizontal_trim(lines: &[&str]) -> Option<usize> {
         let mut i = usize::MAX;
-        let mut can_trim = true;
         let mut first = true;
 
-        for line in &lines {
+        for line in lines {
             for (j, c) in line.chars().enumerate() {
                 if j > i || !"* \t".contains(c) {
-                    can_trim = false;
-                    break;
+                    return None;
                 }
                 if c == '*' {
                     if first {
                         i = j;
                         first = false;
                     } else if i != j {
-                        can_trim = false;
+                        return None;
                     }
                     break;
                 }
             }
             if i >= line.len() {
-                can_trim = false;
-            }
-            if !can_trim {
-                break;
+                return None;
             }
         }
+        Some(i)
+    }
 
-        if can_trim {
-            lines.iter().map(|line| (&line[i + 1..line.len()]).to_string()).collect()
-        } else {
-            lines
+    /// remove a "[ \t]*\*" block from each line, if possible
+    fn horizontal_trim(lines: &mut [&str], i: usize) {
+        for line in lines.iter_mut() {
+            *line = &line[i + 1..];
         }
     }
```

Comment:
It might be simpler just to inline this and remove the function? It's only called once and three lines.

Reply:
I'll add comments then.